### PR TITLE
Back end validation

### PR DESCRIPTION
**Pull request by m4v15**
I think done?

relates #75 #67

- [x] add validation middleware to the backend. This checks each field, and if they don't pass, throws and error
- [x] add error handling middleware.
  - [x] one deals with validation errors - sends a 400 along with the error's back to front end
  - [x] one is generic, sends a 500 server error back
- [x] add 404 route at the bottom of the server routes (this could potentially be made nicer in the future, could serve a static file
- [x] tests for validation (should add some to check the other error handlers)

At the moment, all fields are just set to required, and email is also set to have to be a valid email.

I think this is all that needs doing with the back end. Next, need to change front end to deal with these errors which are sent back as a 400, could do alongside the front end validation as will probably display them in the same way.

### CODE REVIEW

File: src/server/app.js [hunk 1]
```diff
@@ -3,13 +3,36 @@ const path = require("path");
...
 app.use(express.static(path.join(__dirname, "../../public")));
 app.use(bodyParser.json());
 app.use(express.static("public"));
 
+// API Router
 app.use("/api/v1/", api_router);
 
+// Validation Error Handler
+app.use((err, req, res, next) => {
+  if ((err.message = "validation error")) {
```
Comment (ivan-unfolds) [inline src/server/app.js]: Is this meant to be `===`?

**Review by ivan-unfolds — CHANGES_REQUESTED** [review]
Looks good, just one little comment

File: src/server/app.js (hunk 1 above)
Comment (m4v15) [reply src/server/app.js]: Nice spot. Will change this and add a test for it.

**Comment by m4v15** [comment]
@ivanmauricio Spent an hour trying to fake a server error to add a test which would have caught this, could do it but it was quite hacky, so will have it as something to add in later.

**Review by ivan-unfolds — APPROVED** [review]
Looking good!

**Pull request merged**